ground/src/main/java/com/google/android/ground/model/geometry/GeometryValidator.kt

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.ground.model.geometry
+
+object GeometryValidator {
+  private fun List<Any>.isFirstAndLastSame(): Boolean = firstOrNull() == lastOrNull()
+
+  /** Validates that the current [LinearRing] is well-formed. */
+  fun LinearRing.validate() {
+    if (coordinates.isEmpty()) {
+      return
+    }
+    if (!coordinates.isFirstAndLastSame()) {
+      error("Invalid linear ring")
+    }
+  }
+
+  /** Returns true if the current geometry is closed. */
+  fun Geometry?.isClosedGeometry(): Boolean = this is Polygon || this is LinearRing
```

Comment:
Let's combine this condition with the `isComplete`:

```kotlin
fun Geometry.isClosed: Boolean = 
  when(this) {
    is Polygon, LinearRing -> size >= 4 && coordinates.firstOrNull() == coordinates.lastOrNull()
    else -> false
  }
```

This way, in the fragment etc. you can just write:

```kotlin
getButton(ButtonAction.ADD_POINT).showIfTrue(!isClosed)
```

Reply:
The idea behind validation was that it will fail the creation of geometry. So, just the type of geometry is enough to assert whether it is a closed or open geometry. The above `isClosed()` is for `Geometry` and `isComplete()` is a helper method on `List<Coordinate>`.